### ISSUE TITLE
Methods that return back a templated type cause unknown type warnings

### ISSUE THREAD

**Issue by iteriani** [google/closure-compiler#2654]
Repro (using core externs)

Raw JS Module Parent
```
goog.module('t');

/**
 * @param {T} t
 * @return {T}
 * @template T
 */
function bar(t) {return t;}

let C = class C {};
C = bar(C);
exports.C = C;
exports.bar = bar;
```

Raw JS Module Child

```
goog.module('a');

const {bar, C} = goog.require('t');

exports = bar(class A extends C {});
```

This triggers
input1: WARNING - Bad type annotation. Unknown type module$exports$t.C

Presumably because the bar() call causes the type of C to become unknown

**Comment by MatrixFrog** [comment]
https://closure-compiler-debugger.appspot.com/#input0%3Dgoog.module('t')%253B%250A%250A%252F**%250A%2520*%2520%2540param%2520%257BT%257D%2520t%250A%2520*%2520%2540return%2520%257BT%257D%250A%2520*%2520%2540template%2520T%250A%2520*%252F%250Afunction%2520bar(t)%2520%257Breturn%2520t%253B%257D%250A%250Alet%2520C%2520%253D%2520class%2520C%2520%257B%257D%253B%250AC%2520%253D%2520bar(C)%253B%250Aexports.C%2520%253D%2520C%253B%250Aexports.bar%2520%253D%2520bar%253B%26input1%3Dgoog.module('a')%253B%250A%250Aconst%2520%257Bbar%252C%2520C%257D%2520%253D%2520goog.require('t')%253B%250A%250Aexports%2520%253D%2520bar(class%2520A%2520extends%2520C%2520%257B%257D)%253B%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26MISSING_PROPERTIES%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1

**Comment by lauraharker** [comment]
Closing because the [repro](https://closure-compiler-debugger.appspot.com/#input0%3Dgoog.module('t')%253B%250A%250A%252F**%250A%2520*%2520%2540param%2520%257BT%257D%2520t%250A%2520*%2520%2540return%2520%257BT%257D%250A%2520*%2520%2540template%2520T%250A%2520*%252F%250Afunction%2520bar(t)%2520%257Breturn%2520t%253B%257D%250A%250Alet%2520C%2520%253D%2520class%2520C%2520%257B%257D%253B%250AC%2520%253D%2520bar(C)%253B%250Aexports.C%2520%253D%2520C%253B%250Aexports.bar%2520%253D%2520bar%253B%26input1%3Dgoog.module('a')%253B%250A%250Aconst%2520%257Bbar%252C%2520C%257D%2520%253D%2520goog.require('t')%253B%250A%250Aexports%2520%253D%2520bar(class%2520A%2520extends%2520C%2520%257B%257D)%253B%26conformanceConfig%26externs%26refasterjs-template%26includeDefaultExterns%3D1%26CHECK_SYMBOLS%3D1%26MISSING_PROPERTIES%3D1%26TRANSPILE%3D1%26CHECK_TYPES%3D1%26CLOSURE_PASS%3D1%26PRESERVE_TYPE_ANNOTATIONS%3D1%26PRETTY_PRINT%3D1) no longer reports a warning.

Please reopen if you see this again.
